### ISSUE TITLE
group_map() - keep grouping columns

### ISSUE THREAD

**Issue by coolbutuseless** [tidyverse/dplyr#4066]
I know this is late to the party, but I thought I'd request different behaviour for `group_map()` and `group_walk()` while still at the 0.8.0 RC stage.

Currently `group_map()` and `group_walk()` split the data.frame into groups - the data.frame for each group is (a) ungrouped, and (b) does not contain the grouping columns.

I was wondering if there was any thought to having the split data.frame for each group *keeping* the grouping columns?

I found that @hadley wants to think of `group_map()` as a shortcut for **nest-mutate-unnest** (as discussed here - https://github.com/tidyverse/dplyr/pull/3964#discussion_r234686798).

i.e. 

```r
mtcars %>%
  group_by(cyl) %>%
  group_map(~ data.frame(n = nrow(.x)))
```

should be mostly equivalent to

```r
mtcars %>% 
  group_by(cyl) %>% 
  nest() %>% 
  mutate(out = map(data, ~ data.frame(n = nrow(.x)))) %>% 
  unnest(out)
```

However, treating `group_map` as a shortcut for **split-apply-combine** makes a similar amount of sense (and has the added benefit that `group_walk` can be easily described as a shortcut for **split-apply** without the *combine* step). 

If we go along with this line of thinking then @hadley's example in `group_map` would be equivalent to a `purrr::map_dfr`, i.e.

```r
mtcars %>% 
  group_by(cyl) %>% 
  group_split() %>%
  map_dfr(data.frame(n = nrow(.x))) 
```

(NB - the above code does not include the extra code needed to re-apply the grouping of the original data).

My issue with trying to mimic **nest-mutate-unnest** with `group_map` is that the `map` nomenclature doesn't make it readily apparent that you're not getting the full data for each group in the split data.frames  (whereas when using `nest()` we are explicitly asking for this hierarchy of *data.frame-within-group*, and understand we are asking only for the data nested within the grouping level)

E.g. My initial use for `group_walk()` was going to be something like:

```r
mtcars %>%
  group_by(cyl) %>%
  group_walk(~write.csv(.x, paste0(.x$cyl[1], ".csv")))
```

But this won't be correct, as the grouping columns don't exist in `.x`, and I would not be saving the full data within the `walk()`.

Are there cases where having the grouping columns present in the split data.frames would be a problem?  

Are there other reasons for choosing to drop grouping columns from the split data.frame? (I went back through GH issues but couldn't find any justification other than the suggestion to try and be equivalent to nest-mutate-unnest).

**Comment by hadley** [comment]
The motivation for not including them is that you're not responsible for including them in the output, which makes most computation easier. Your purrr code is a good example:

``` r
library(dplyr)
library(purrr)

mtcars %>% 
  group_by(cyl) %>% 
  group_split() %>%
  map_dfr(~ data.frame(n = nrow(.x))) 
#>    n
#> 1 11
#> 2  7
#> 3 14
```

That's not useful because you've lost the grouping information, and it seems unappealing to require people to include the grouping variables in every output. With the current system, you can access the grouping information if you need it, via the secondary argument to the mapping function.

The argument is less strong for `group_walk()` since the output is discarded, but it seems unappealing for `group_map()` and `group_walk()` to have different interfaces. That said, we might consider adding an argument to control this behaviour in future version.

**Comment by romainfrancois** [comment]
@coolbutuseless you do this instead: 

```r
mtcars %>%
    group_by(cyl) %>%
    group_walk(~write.csv(.x, paste0(.y$cyl, ".csv")))
```

i.e. use `.y$cyl` and you don't need the `[1]` here because `.y` is always 1 row, no matter what, even if `.x` has 0 rows. 


It would not be too hard to have a `keep=` argument in `group_map()` that would just go to `group_split()` so that `.x` also has the grouping variables but: 

 - it makes things more difficult on the way out
 - when `.x` has 0 rows, `.x$cyl` would not be useful where as `.y$cyl` is always one value

**Comment by coolbutuseless** [comment]
Thanks for the comments @hadley , @romainfrancois .  

I fully appreciate the difficult design/programming choice here.

Choice 1: keep a clean division -  with the grouped + ungrouped values appearing in 2 separate data.frames()  - but now with the overhead of having to keep track of which data.frame each variable is in (and having to re-join them as necessary)

Choice 2: keep all the data as a single data.frame "slice" of the original data, but then this adds the  overhead of "How do you let the user know what the group/slice variables were?", which is difficult(?) to do well and have a consistent interface across all the other group/ungroup ideas in dplyr.

Either choice makes it difficult for those that handle data in the other way.  

As it stands, "group_split()" + "map_dfr()" do *exactly* what i need (and does it in the way I reason about my data).


In summary: Thanks for "group_split()" :)

**Comment by hadley** [comment]
See https://twitter.com/coolbutuseless/status/1096983675013554176 for an example of where an argument would be nice.

**Comment by romainfrancois** [comment]
adding `keep` in #4251 but also repurposed what `group_map()` does.

**Comment by romainfrancois** [comment]
Done in #4251
